### ISSUE TITLE
file monitor reportedly unsupported

### ISSUE THREAD

**Issue by uohmak4fvpqe** [gittup/tup#434]
I cannot use ```tup monitor``` anymore :
```
> uname -a
Linux laptop 5.13.13-1070.native #1 SMP Thu Aug 26 08:03:35 PDT 2021 x86_64 GNU/Linux
> tup -v
tup 0.7.11
> tup monitor
tup error: The file monitor is not supported on this platform.
```
It used to work fine before I reinstalled my system (ClearLinux). I don't know if a package is missing. I'm also using a ```luks``` partition now, hoping it's not the issue. ```tup``` was installed via ```brew```.

```inotify``` itself seems to work after ```swupd bundle-add inotify-tools``` : 
```
> inotifywait . # then delete file 'test'
Setting up watches.
Watches established.
./ DELETE test
```
So I think it should be possible for tup monitor to work

**Comment by gittup** [comment]
Did you install tup from a package in ClearLinux, or build it yourself from the tup repo? The bootstrap.sh builds a temporary version of tup that doesn't include the monitor which is then used to bootstrap the final tup with the monitor installed. So I'm wondering if you're using a package that maybe packaged the bootstrap version instead of the final build.

Probably the best fix is to have bootstrap.sh match the build description in the Tupfiles more accurately if this is the case.

**Comment by uohmak4fvpqe** [comment]
As mentionned, I installed tup from ```brew```. I see there is no file named ```bootstrap.sh``` in ```~/.linuxbrew```

**Comment by uohmak4fvpqe** [comment]
I build it. Monitor or not, It is now unable to build my projects... I can't be very specific, it seems to think that the same tupfiles that work with brew's version are wrong

**Comment by gittup** [comment]
> As mentionned, I installed tup from `brew`. I see there is no file named `bootstrap.sh` in `~/.linuxbrew`
> Now you mention, I believe I actually built it myself last time. There is no version bundled bytheway

Ahh sorry I missed that. It looks like the tup formula just runs build.sh, and uses the intermediate version of tup that is built without the monitor. I'll try to fix that so build.sh produces a fully useable tup.

> I build it. Monitor or not, It is now unable to build my projects... I can't be very specific, it seems to think that the same tupfiles that work with brew's version are wrong

That is very concerning. What error message are you hitting? Can you post a reproducible example?

**Comment by uohmak4fvpqe** [comment]
[here a simplified one](https://file.io/1gpTP1zR0sTM)
the included file ```./tup``` is my own compiled version.

note : both ```tup``` and ```./tup``` work for me;
however using variant : only ```tup variant conf/*; tup``` do, ```./tup variant conf/*; ./tup``` doesn't

**Comment by gittup** [comment]
> [here a simplified one](https://filein.io/Cx38M_IrKF.xz)
> the included file `./tup` is my own compiled version.
> 
> note : both `tup` and `./tup` work for me;
> however using variant : only `tup variant conf/*; tup` do, `./tup variant conf/*; ./tup` doesn't
> 
> you'll need clang and libc++ to reproduce

Thanks! What versions of clang and libc++ are you using? My default version of clang (10.0.0-4ubuntu1) is complaining about some of the -W flags, so I want to make sure I'm accurately reproducing what you're seeing.

**Comment by gittup** [comment]
Also, I ended up pushing a fix so that the bootstrapped version will include the file monitor for Linux. So the next release in homebrew should include the monitor, but hopefully I can include a fix for the other issue as well!

**Comment by uohmak4fvpqe** [comment]
Thanks a lot about the monitor

so :
- ```clang``` version is ```11.1.0```
- ```_LIBCPP_VERSION``` has value ```10000``` (ony way I found, clearlinux has no distinct packages for stuff...)

**Comment by gittup** [comment]
Thanks, I think I was able to reproduce the issue. You're seeing a failure like this, right?

```
[ tup ] [0.048s] Executing Commands...
 2) [0.050s] [build-native] src/cc/dummy_module: ../../../build-native/src/cc/dummy_module/helloworld.pcm                                                    
* 1) [build-native] src/cc/dummy_module: ../../../build-native/src/cc/dummy_module/test.cc.o                                                                 
test.cc:2:8: fatal error: module 'helloworld' not found
import helloworld;
~~~~~~~^~~~~~~~~~
```

I believe the root problem here is just the changes to variants. With v0.7.11 and prior variants worked by doing a filesystem overlay, so build-native/src/cc/dummy_module and src/cc/dummy_module appeared as a single directory to the subprocess (clang in this case). For the most part this meant you could write Tupfiles as if variants didn't exist, and then use a variant without changing Tupfiles. However, there were some irreconcilable issues with that approach, the first being that it relied on FUSE making multiplatform support very difficult, and the second being that the filesystem overlay doesn't exist outside of tup, so anything that stored the path in a generated file (like gdb) would likely be pointing to the wrong directory.

The new variants unfortunately make you sometimes point to both the src directory and the variant directory. In this case, I think you just need to point the -fprebuild-module-path to the variant directory. This change seems to work for me:

```
diff --git a/src/cc/dummy_module/Tupfile b/src/cc/dummy_module/Tupfile
index f0fe651..7133ca5 100644
--- a/src/cc/dummy_module/Tupfile
+++ b/src/cc/dummy_module/Tupfile
@@ -1 +1,2 @@
+cc_flags_include += -fprebuilt-module-path=$(TUP_VARIANTDIR)
 include_rules
```

Unfortunately putting the flag in your script/tuprules_clang++.tup file doesn't work since it points to the build-native/script/ directory instead of the build-native/src/cc/dummy_module/ directory. (ie: it points to where the included tuprules_clang++.tup file is, not where the Tupfile is). There should probably be another special tup variable that points to the subdirectory to help facilitate this.

**Comment by uohmak4fvpqe** [comment]
Okay so it was actually an breaking update. Frankly I didn't realize the project was that young

Thanks for the explanations anyway. I'm on something else these days but I will certainly make use of it

**Comment by gittup** [comment]
> Okay so it was actually an breaking update. Frankly I didn't realize the project was that young

Yes; this was announced back in May: https://groups.google.com/g/tup-users/c/Fw5L-ii1VnI

And it will be announced again during the next release. This is probably one of the only intentional breaking updates in the history of tup, and the fact that it would be breaking is one of the reasons that I delayed in implementing it for so long (which unfortunately meant more people are using it). I find it frustrating when my work is broken by other project updates, so I try very hard not to inflict that on my users.

Sorry for the inconvenience.

**Comment by uohmak4fvpqe** [comment]
Alright so your solution works. However it is indeed a real pain to not be able to refer to (the variant version of) the place were the macro is *used* from. If there was such a variable I could factorize my code as before.
